File: 141-linked-list-cycle/linked-list-cycle.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(h *ListNode) bool {
    if h == nil || h.Next == nil {
        return false
    }
    
    slow := h
    fast := h
    
    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
        if slow == fast {
            return true
        }
    }
    return false
}

File: 141-linked-list-cycle/linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *h) {
        if (h == nullptr || h->next == nullptr) {
            return false;
        }
        
        ListNode* s = h;
        ListNode* f = h;
        
        while (f != nullptr && f->next != nullptr) {
            f = f->next->next;
            s = s->next;
            if (f == s) {
                return true;
            }
        }
        return false;
    }
};
